cloudinit/net/dhcp.py

```diff
@@ -654,7 +654,7 @@ def dhcp_discovery(
             if is_ib_interface(interface):
                 infiniband_argument = ["--clientid"]
             command = [
-                self.dhcp_client_path,  # pyright: ignore
+                self.client_name,
```

Comment:
Would it be better to update the constructor so that the path can only have a string? 
E.g., instead of
```python
    def __init__(self):
        self.dhcp_client_path = subp.which(self.client_name)
        if not self.dhcp_client_path:
            raise NoDHCPLeaseMissingDhclientError()
```
do something like
```python
    def __init__(self):
        dhcp_client_path = subp.which(self.client_name)
        if not dhcp_client_path:
            raise NoDHCPLeaseMissingDhclientError()
        self.dhcp_client_path = dhcp_client_path
```

Reply:
> Would it be better to update the constructor so that the path can only have a string?

I'm fine with checking in `__init__()`, but passing absolute paths to `subp` isn't necessary and makes the logged commands more noisy to read.